core/primitives/src/bandwidth_scheduler.rs

```diff
@@ -69,8 +73,177 @@ pub struct BandwidthRequestsV1 {
+        }
+
+        // The value that is closest to max_receipt_size is set to max_receipt_size.
+        // Without this we could end up in a situation where one shard wants to send
+        // a maximum size receipt to another and requests the corresponding value from the list,
+        // but the sum of this value and base bandwidth exceeds max_shard_bandwidth.
+        // There's a guarantee that num_shards*base_bandwidth + max_receipt_size <= max_shard_bandwidth,
+        // but there's no guarantee that num_shards*base_bandwidth + request_value <= max_shard_banwdidth.
+        let mut closest_to_max: Bandwidth = 0;
+        for value in &values {
+            if value.abs_diff(params.max_receipt_size)
+                < closest_to_max.abs_diff(params.max_receipt_size)
+            {
+                closest_to_max = *value;
+            }
+        }
+        for value in values.iter_mut() {
+            if *value == closest_to_max {
+                *value = params.max_receipt_size;
+            }
+        }
+
+        BandwidthRequestValues { values }
+    }
+}
+
+/// Bitmap which describes which values from the predefined list are being requested.
+/// The nth bit is set to 1 when the nth value from the list is being requested.
+#[derive(
+    BorshSerialize,
+    BorshDeserialize,
+    serde::Serialize,
+    serde::Deserialize,
+    Clone,
+    PartialEq,
+    Eq,
+    ProtocolSchema,
+)]
+pub struct BandwidthRequestBitmap {
+    pub data: [u8; BANDWIDTH_REQUEST_BITMAP_SIZE],
+}
+
+impl std::fmt::Debug for BandwidthRequestBitmap {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "BandwidthRequestBitmap(")?;
+        for i in 0..self.len() {
+            if self.get_bit(i) {
+                write!(f, "1")?;
+            } else {
+                write!(f, "0")?;
+            }
+        }
+        write!(f, ")")
+    }
+}
+
+pub const BANDWIDTH_REQUEST_BITMAP_SIZE: usize = BANDWIDTH_REQUEST_VALUES_NUM / 8;
+const _: () = assert!(
+    BANDWIDTH_REQUEST_VALUES_NUM % 8 == 0,
+    "Every bit in the bitmap should be used. It's wasteful to have unused bits.
+    And having unused bits would require extra validation logic"
+);
+
+impl BandwidthRequestBitmap {
+    pub fn new() -> BandwidthRequestBitmap {
+        BandwidthRequestBitmap { data: [0u8; BANDWIDTH_REQUEST_BITMAP_SIZE] }
+    }
+
+    pub fn get_bit(&self, idx: usize) -> bool {
+        assert!(idx < self.len());
+
+        (self.data[idx / 8] >> (idx % 8)) & 1 == 1
+    }
+
+    pub fn set_bit(&mut self, idx: usize, val: bool) {
+        assert!(idx < self.len());
+
+        if val {
+            self.data[idx / 8] |= 1 << (idx % 8);
+        } else {
+            self.data[idx / 8] &= !(1 << (idx % 8));
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        BANDWIDTH_REQUEST_VALUES_NUM
+    }
+
+    pub fn is_all_zeros(&self) -> bool {
+        self.data == [0u8; BANDWIDTH_REQUEST_BITMAP_SIZE]
+    }
```

Comment:
I must say that implementing my own bitmap from scratch feels a bit wrong. I should probably use an existing library like `bitvec`.
The problem with `bitvec` is that I want something that:
* Is a mutable bitmap
* Is borsh serializable/deserializable

The problem with bitvec is that it's not borsh serializable.
The closest thing I could find in existing code is `ChunkEndorsementsBitmap`, but it uses `bitvec` to convert from a list of bools to raw bytes. It's not the pretties solution. Having raw bytes (`Vec<u8>`) is a bit unreadable, and it requires the person who creates the struct to already know what bits should be set in advance, it's not mutable.
Another option would be to convert from Vec<bool> to bits druing serializatoin/deserialization, but I don't want to get into that.

IMO having a dedicated bitmap struct is the most readable option. Maybe there should be a generic bitmap struct that can be used in protocol structs?

Reply:
Can you ask on out zulip in the rust channel?

Reply:
I read a bit more into the `bitvec` crate and I found a neat thing - https://docs.rs/bitvec/latest/bitvec/slice/struct.BitSlice.html.
I think it will allow me to view the array as a list of individual bits, without having to manipulate bits myself.